pilot/pkg/security/authz/builder/extauthz.go

```diff
@@ -181,13 +194,23 @@ func parseService(service string, port int) (hostname string, cluster string, er
 	if len(parts) == 2 {
 		namespace, name = parts[0], parts[1]
 	} else {
-		// TODO(yangminzhu): Fix to use the namespace of MeshConfig.
-		namespace, name = "istio-system", parts[0]
+		namespace, name = in.Push.MeshConfigNamespace, parts[0]
+	}
+
+	for _, svc := range in.Push.Services(in.Node) {
```

Comment:
I have concerns about scale here, previous loops on service caused outages due to poor performance

Reply:
Really, I don't think we should be using service name at all. Its not consistent with how anything else works, including mesh config "cluster local" setting, which uses hostname. We will have issues referring to ServiceEntry (which the name is the hostname) and potential conflicts. Given virtually every API in istio is host name based, I don't think changing this is a good idea.

Reply:
Ah, I realized you added ResourceName - didn't see that at the time of my comment. I still agree with what I said before, I think adding semantic meaning to the ServiceEntry name when it had none before is a mistake.

Reply:
`in.Push.Services()` should handle the "cluster local" setting automatically for us, what we get here should be the final result of whatever setting used in mesh config. and the conflict resolution is just to pick the first available hostname when multiple are found in ServiceEntry.

For the other 2 issues:
1. scale of looping on services: can you give some more details about the concern, we have multiple places looping over the services (e.g. buildOutboundClusters, DestinationRule, UpdateServiceShards)

2. service convention: my understanding is that there is a trend to use the resource name to refer to another service, example: the HTTPProxy resource in the https://www.cncf.io/blog/2020/10/29/kubernetes-1-19-the-future-of-traffic-ingress-and-routing/, we also have the delegate VirtualService that uses the name of another VS right? Even it is for a different purpose but I think it's not very uncommon to refer to another resource by its name. The last thing I heard about hostname is that it exposes the DNS resolution detail to users and the service name actually help to hide that as an additional indirection. I think that's what we agreed from the meeting to finalize the API design for this, cc @louiscryan @smawson @costinm if we want to revisit this decision.

Reply:
HTTPProxy is a Countour thing, not a standard. The https://github.com/kubernetes-sigs/service-apis does have references to service names, but its not directly applicable here, because they both allow setting an explicit type (so they user would say, "This is a service entry named foo") rather than magically selecting between two, and they don't have cross-namespace.

For the scale: we loop through those 1x, but now for each service that has ext_authz we loop through all services again, so its now n^2 I think? https://github.com/istio/istio/pull/25294 is an example of breaking out a similar looping and seeing 80% pilot CPU reduction. I am more concerned about the API than the perf though.

Reply:
I think they actually both allow cross-namespace reference: https://projectcontour.io/docs/v1.10.0/config/api/#projectcontour.io/v1.Include, https://kubernetes-sigs.github.io/service-apis/spec/#networking.x-k8s.io/v1alpha1.RouteBindingSelector. 

And at least the contour API doesn't have the type field instead it just only allows HTTPProxy, but the https://kubernetes-sigs.github.io/service-apis/spec/#networking.x-k8s.io/v1alpha1.RouteBindingSelector seems more flexible that has the `group`/`kind` field. 

I agree it is confusing if there are service and serviceentry with the same name, but I think we probably have to live with it at this moment with this experimental API,  of course we will document this clearly and improve the logging for it. I think this we can probably solve this in a later version of the API together with many other issues, like using explicit `name` and `namespace`, putting it in a dedicated CRD, etc. 

Re the scale, currently it's per workload that has CUSTOM action, not per service, so it's still O(n) where n is the number of the service, we can later refactor it to do only once per push (instead of per workload)

Reply:
This looks quite optimizable. Getting the list of all the configured provider name in the policy and then iterating once in all meshconfig's providers. We can do that in the init time when the push context is inited?

Reply:
probably not in init time since the service is dynamic, we can just do it in the beginning of push, should be good enough already.

Reply:
RouteBindingSelector is not selecting a service, and its a label match not a direct reference. The closer thing is https://kubernetes-sigs.github.io/service-apis/spec/#networking.x-k8s.io/v1alpha1.HTTPRouteForwardTo, which is directly referencing a service or a generic k8s object with https://kubernetes-sigs.github.io/service-apis/spec/#networking.x-k8s.io/v1alpha1.LocalObjectReference.

However, just because they have an API to reference objects does not inherently mean that it makes sense to make ServiceEntry a reference-able object. There are a huge number of ServiceEntries which are not possible to forward traffic to.

For example,

```
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  name: external
spec:
  hosts:
  - foo.org
  - bar.org
  location: MESH_EXTERNAL
  ports:
  - number: 80
    name: tcp
    protocol: TCP
  resolution: NONE
```

This will create two clusters, so now we are arbitrarily choosing one. It also is an ORIGINAL_DST cluster, which I am almost certain envoy would reject

Reply:
we won't support any ServiceEntry, at least not for now. ORIGINAL_DST will not be supported regardless of the reference method, it is not supported even if we specify host name directly.

Multiple host names is okay if they have the same functionality but is not the recommend way to use with ext-authz, we will recommend the user to create a dedicated ServiceEntry with just 1 host name to use with ext-authz.

Reply:
My concern isn't that its impossible to reference a valid ServiceEntry, my concern is there are 10 ways to reference an invalid ServiceEntry. We shouldn't make it so easy to do the wrong the in the API.

I think there is a reason the "ResourceName" attribute is not a part of model.Service yet - its not the right entrypoint for referencing a service

Reply:
discussed offline and changed to use hostname, also remove the ResourceName/Namespace as it's no longer needed.